### ISSUE TITLE
[Tiled] Make tiles in a layer accessible 

### ISSUE THREAD

**Issue by musbah** [craftworkgames/MonoGame.Extended#265]
Hello, I've recently started using this library and I though I would try to contribute (even though I can't really call this much). Anyway I was wondering if it would be a good idea to make a getter for the _tiles variable in TiledTileLayer.cs .

 I've been using the tiled collision editor in my project and noticed that I could get all the tileset objects easily but that I couldn't actually "attach" the objects to the tiles on the map (basically get the tile information to set the positions and so on) because the variable was set to private. So I just created a getter method for my own project and I though that the idea could maybe benefit someone else.

I didn't create a pull request because I wasn't sure if it should be added (there could be something I'm overlooking) and there might be a more efficient way to do this instead of just creating a getter.

**Comment by craftworkgames** [comment]
Could you post the snippet of code you changed / want changed?

**Comment by musbah** [comment]
What I did was add the following code in TiledTileLayer.cs

`        public TiledTile[] GetTiles()
        {
            return _tiles;
        }`

and then I used `((TiledTileLayer)_tiledMap.GetLayer("solids")).GetTiles();` to get the tiles.
There might be a better way of doing this that I just couldn't think of though.

**Comment by OptimusPi** [comment]
You should be using objects for collisions, yeah. Look at Demos.Platformer. Tiles are simply visual. Objects in Tiled Map Editor are loaded with the service as Entities and work in the whole entity system very nicely.

**Comment by musbah** [comment]
The thing is when it comes to the object layer it's annoying to recreate the same polyline or whatever over and over again, yes you can copy it but you'll still have to move it and put it on every single tile in the layer. If you use the tiled collision editor you give the object to the tileset tile itself, you therefore don't need an object layer, the tileset itself will contain the object. The problem with that is that you'll need to compare the tileset tile objects to the tiles on the map and that's what I'm trying to do above.

**Comment by craftworkgames** [comment]
Yeah, there's definitely more than one way to solve these kinds of problems. The platformer demo uses objects for collisions but tbh it's not a particularly scalable solution. For small games it's fine, but with a little more work I agree with @musbah that using the Tiled collision editor is a better way to go.

Put simply, I think the goal here is to enable developers to make the games the way they want to make them. It's not always easy, we can't possibly cater for every situation, but we can certainly do better.

I've been meaning to get back to this issue over the last couple of days. Unfortunately, I haven't been feeling great and I've been super busy at my day job. 

I think what we should do is go through the API and take a look at the things that haven't been exposed. In most cases, I think it makes sense to expose lists as properties using the `IReadOnlyList` interface much like we've done in the [TiledMap.cs](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended/Maps/Tiled/TiledMap.cs) file.

        public IReadOnlyList<TiledTileset> Tilesets => _tilesets;
        public IReadOnlyList<TiledObjectGroup> ObjectGroups => _objectGroups;
        public IReadOnlyList<TiledLayer> Layers => _layers;

This approach gives us the best of both worlds. The data is exposed as an indexable list but still protected so that the state of the map remains stable. Using properties is better than using a method like `GetTiles` because it plays nicely with intellisense and follows C# guidelines.

Unfortunately, there's been a number of PR's merged into the Tiled code that has created some inconsistencies in the API. It could really do with some cleanup.

**Comment by musbah** [comment]
That sounds like a good plan, I'll try to look through the API and find some of the inconsistencies and maybe record them somewhere for comparison.

**Comment by craftworkgames** [comment]
> I'll try to look through the API and find some of the inconsistencies and maybe record them somewhere for comparison.

Excellent idea.

**Comment by musbah** [comment]
I just noticed that [TiledTileLayer.cs](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended/Maps/Tiled/TiledTileLayer.cs) already had a way to get the tiles, I just never saw it.

`public IEnumerable<TiledTile> Tiles => _tiles;`

Could `IEnumerable` be better than `IReadOnlyList` in this case? I guess it's assuming that we only need to look at each tile once, which makes sense.

**Comment by craftworkgames** [comment]
The main reason we used `IEnumerable` is because `IReadOnlyList` was only introduced in .NET 4.5 and we used to target .NET 4.0.

I think we should use `IReadOnlyList` in pretty much every case from now on.

**Comment by musbah** [comment]
Should we actually go through everything that uses `IEnumerable` and change them to `IReadOnlyList` then? Or is that just a waste of time? I can't really find a lot of information when looking up the differences between them.

**Comment by musbah** [comment]
Ok, I more or less went through all the classes inside of the tiled folder and there seems to be 2 types of ways to show the results to the people users : 
`public List<TiledObjectGroup> ObjectGroups { get; private set; }`
or
 `public IReadOnlyList<TiledObjectGroup> ObjectGroups => _objectGroups;`

**Comment by craftworkgames** [comment]
> Should I go through them and make them use the 2nd way? Just to kind of make everything look more or less the consistent.

Yes please.

Exposing a `List<TiledObjectGroup>` directly is generally a bad idea. It limits what we can do behind the scenes in terms of optimizations.

**Comment by musbah** [comment]
I've looked at everything even closer, and it seems that changing them to an `IReadOnlyList` will only cause problems to the other pieces of code in the tiled folder that use the values.

When it comes to [TiledTileSet.cs](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended/Maps/Tiled/TiledTileSet.cs) I noticed that changing : 
`public List<TiledTileSetTile> Tiles { get; private set; }` is not a good idea, because it's used by TileMap.cs and TiledTileLayer.cs so I didn't touch it.

Something similar happens when changing [TiledTileSetTile.cs](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended/Maps/Tiled/TiledTileSetTile.cs) as well.

We would probably need to either keep them as is, or try to find some other solution. Now to be fair they're not all necessarily needed by the end user, but some things seem like they would be useful to have.

**Comment by LithiumToast** [comment]
@musbah 
>I've looked at everything even closer, and it seems that changing them to an IReadOnlyList will only cause problems to the other pieces of code in the tiled folder that use the values.

What's the problem exactly? `List<T>` implements `IReadOnlyList<T>` and `IEnumerable<T>`. `IReadOnlyList<T>` prevents the list from being mutated while basically allowing it to be iterated still using a for loop and a for each loop. `IEnumerable<T>` basically does the same, except it only allows iteration using a foreach loop. Using `IReadOnlyList<T>` is probably the best in all scenarios if the list shouldn't change state at runtime which is what we want I think.

**Comment by rafaelalmeidatk** [comment]
Sorry my ignorance, maybe I haven't understood out your problem, but don't we already have access of the tiles on a given layer?  I'm saying that because in my game we use the tiles on a specif layer to create the collisions, this way we don't have to create an object to every collision. Here is the method: https://github.com/rafaelalmeidatk/Super-Pete-The-Pirate/blob/master/Super%20Pete%20The%20Pirate/Objects/GameMap.cs#L90

**Comment by craftworkgames** [comment]
> We would probably need to either keep them as is, or try to find some other solution.

I think it's worth at least considering what it would take to make the API consistent. 

There's actually quite a number of things I'd like to do to improve the Tiled API since we started working on the new renderer. This is just one of them. When I get a chance I'll sit down and take a closer look.

**Comment by musbah** [comment]
@LithiumToast I actually tried implementing them on my own version and the error I got was that TileMap and TiledTileLayer did not have enough permission, so I assume that they do mutate the list (I didn't go through the code yet, a bit busy at the moment). I could have just implemented it incorrectly though, that's possible.

@rafaelalmeidatk You are correct, after creating the issue I did notice that what I wanted actually did exist, but this topic kind of evolved and so I forgot all about that. Thank you for reminding me about it though and also thank you for the example, your code looks way better than mine.

@craftworkgames  I think we should probably close this issue and create another one for the Tiled API cleanup, this thread kind of went all over the place.
